DCE inference graphs too

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97275

I added a bunch of asserts to verify that I didn't accidentally kill copy_ in the graph, hopefully this combined with our existing tests is good enough.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
DCE inference graphs too

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1283,11 +1287,27 @@ def aot_dispatch_base(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig, *
+    copy_count = 0
+    # NB: It would also be nice to verify that the mutations all happen at the
+    # end, but we also do some administrative views after mutations so this
+    # isn't actually true.  (TODO: Could this cause problems for Inductor?)
```
Hmm, I vaguely remember this being a problem, but I thought we "fixed" it by syncing both the inputs and outputs before doing any copies during tracing. But I might be wrong (do you know of an instance where this happens?).

ezyang replied:
yeah it failed on `pytest test/functorch/test_aotdispatch.py -k test_input_output_view_mutate_multiple`

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1258,12 +1258,16 @@ def aot_dispatch_base(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig, *
+
+    copy_count2 = assert_functional_graph(fw_module.graph, allow_input_mutations=aot_config.keep_inference_input_mutations)
+
+    assert copy_count == copy_count2
```
seems very reasonable

bdhirsh reviewed (approved):
lgtm!

ezyang commented:
@pytorchbot merge -f "unrelated problems"

ezyang commented:
@pytorchbot revert -c ignoredsignal -m "this broke a test"

SherlockNoMad commented:
@pytorchbot revert -m "breaking inductor test"

ezyang committed:
Update on "DCE inference graphs too"


I added a bunch of asserts to verify that I didn't accidentally kill copy_ in the graph, hopefully this combined with our existing tests is good enough.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
